src/ripple/app/tx/impl/ApplyContext.cpp

```diff
@@ -69,4 +71,55 @@ ApplyContext::visit (std::function <void (
     view_->visit(base_, func);
 }
 
+template<std::size_t... Is>
+TER
+ApplyContext::checkInvariantsHelper(TER terResult, std::index_sequence<Is...>)
+{
+    if (view_->rules().enabled(featureEnforceInvariants))
+    {
+        auto checkers = getInvariantChecks();
+
+        // call each check's per-entry method
+        visit (
+            [&checkers](
+                uint256 const& index,
+                bool isDelete,
+                std::shared_ptr <SLE const> const& before,
+                std::shared_ptr <SLE const> const& after)
+            {
+                // Sean Parent for_each_argument trick
+                (void)std::array<int, sizeof...(Is)>{
+                    {((std::get<Is>(checkers).
+                            visitEntry(index, isDelete, before, after)), 0)...}
+                };
+            });
+
+        // Sean Parent for_each_argument trick
+        // (a fold expression with `&&` would be really nice here when we move
+        // to C++-17)
+        std::array<bool, sizeof...(Is)> finalizers {{
+            std::get<Is>(checkers).finalize(tx, terResult, journal)...}};
+
+        // call each check's finalizer to see that it passes
+        if (! std::all_of( finalizers.cbegin(), finalizers.cend(),
+                [](auto const& b) { return b; }))
+        {
+            terResult = (terResult == tecINVARIANT_FAILED) ?
+                tefINVARIANT_FAILED :
+                tecINVARIANT_FAILED ;
+            JLOG(journal.error()) <<
+                "Transaction has failed one or more invariants";
```

Comment:
would logging the JSON for the transaction be sufficient, or do we need more? Should I just append it to the current error log message or do we prefer to have a second message at a different log-level (say warn or info) that has the detailed info?